pal/src/misc/random.cpp

```diff
@@ -18,34 +18,47 @@ Module Name:
 --*/
 
 #include "pal/palinternal.h"
-#include <sodium.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <mutex>
 
-/*++
-Function :
-RANDOMInitialize
-
-Initialize libsodium
 
-(no parameters)
+#define RANDOM_CACHE_SIZE 8
+constexpr int rand_byte_len = sizeof(unsigned int) * RANDOM_CACHE_SIZE;
+static union {
+  char rstack[rand_byte_len];
+  unsigned int result[RANDOM_CACHE_SIZE];
+} random_cache;
+static unsigned cache_index(RANDOM_CACHE_SIZE);
+static std::mutex random_generator_mutex;
 
-Return value :
-TRUE  if libsodium initialization succeeded
-FALSE otherwise
---*/
-BOOL RANDOMInitialize(void)
+static bool GetRandom(unsigned int *result) noexcept
 {
-    if (sodium_init() == -1)
-    {
-        return FALSE;
+    std::lock_guard<std::mutex> guard(random_generator_mutex);
+    if (cache_index < RANDOM_CACHE_SIZE) {
+        *result = random_cache.result[cache_index++];
+        return true;
     }
-    
-    return TRUE;
-}
 
+    const int rdev = open("/dev/urandom", O_RDONLY);
+    unsigned len = 0;
+
+    do
+    {
+        int added = read(rdev, random_cache.rstack + len, rand_byte_len - len);
+        if (added < 0) {
```

Comment:
random device driver (random, urandom) is based on an entropy pool. When that pool is empty, it will block the execution until it has something. So, it is either positive or not.